### PR TITLE
Feature/pagination v2

### LINKED ISSUE #28: Management subjects listing endpoint needs pagination
<a href="https://github.com/Jakedavies"></img></a> **Issue by [Jakedavies](https://github.com/Jakedavies)**
_Tuesday May 22, 2018 at 15:22 GMT_
_Originally opened as https://github.com/clevertech/OpenGDPR/issues/229_

----

add a parameter to the subject listing endpoint that adds `?page=1`

### PR DESCRIPTION
**Pull request by sindelio**
Closes #28 
Pagination finished, tested with Postman, all good.
Tests done, but 2 of them require changes to the db setup
Blocked by #68

### CODE REVIEW

**Comment by Jakedavies** [comment]
You need to fix the failing tests you mentioned this morning before we can approve this, I know its also blocked by #68 being ready.

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 1]
```diff
@@ -1,12 +1,48 @@
 const SubjectsService = require('./subjects.service');
 
+const PAGINATION_COUNT = 2; // HARD CODED CONSTANT
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.controller.js]: This variable should be renamed PAGE_SIZE, also 2 is too small

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 2]
```diff
@@ -1,12 +1,48 @@
...
-    res.json(await this.subjectsService.listSubjects());
+    const subjectsList = await this.subjectsService.listSubjects();
+    var requestedPage = req.query.page;
+    var pages = [];
+
+    if (requestedPage === undefined) {
+      requestedPage = 1;
+    }
+
+    for (var i = 0; i * PAGINATION_COUNT < subjectsList.length; i++) {
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.controller.js]: pagination should be done in the DB, take a look at the `limit` and `offset` functions, https://www.petefreitag.com/item/451.cfm

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 3]
```diff
@@ -1,12 +1,48 @@
...
+      pages[i] = subjectsList.slice(i * PAGINATION_COUNT, (i + 1) * PAGINATION_COUNT);
+    }
+
+    if (i == 0) {
+      // This block handles the case in which there are no subjects in the db
+      if (requestedPage == 1) {
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.controller.js]: I don't think this logic is needed if you use DB pagination

File: cg/test/management/subjects.api.spec.js [hunk 4]
```diff
@@ -2,9 +2,13 @@ const { initResources, fetch, closeResources } = require('../utils');
 const { db } = require('../../src/db');
 const { generateClientKey, encryptForStorage, hash } = require('../../src/utils/encryption');
 const { managementJWT } = require('../../src/utils/jwt');
-const { Unauthorized } = require('../../src/utils/errors');
+const { BadRequest, Unauthorized } = require('../../src/utils/errors');
 
 beforeAll(initResources);
+// beforeEach(async () => {
```
Comment (Jakedavies) [inline cg/test/management/subjects.api.spec.js]: remove this

File: cg/test/management/subjects.api.spec.js [hunk 5]
```diff
@@ -194,4 +198,168 @@ describe('List subjects that have given consent', () => {
...
+        message: 'child "page" fails because ["page" must be a number]',
+        validation: {
+          source: 'query',
+          keys: ['page']
+        }
+      })
+    );
+  });
+
+  it('should allow a page query with a valid page number', async () => {
```
Comment (Jakedavies) [inline cg/test/management/subjects.api.spec.js]: We should also be validating that multi-page queries work. Check that you get served a page 2 that contains the expected items too

Comment (sindelio) [reply cg/test/management/subjects.api.spec.js]: Alright

File: cg/test/management/subjects.api.spec.js (hunk 4 above)
Comment (sindelio) [reply cg/test/management/subjects.api.spec.js]: Ok

File: cg/src/domains/management/subjects/subjects.controller.js (hunk 3 above)
Comment (sindelio) [reply cg/src/domains/management/subjects/subjects.controller.js]: Ok I will look into it

File: cg/src/domains/management/subjects/subjects.controller.js (hunk 1 above)
Comment (sindelio) [reply cg/src/domains/management/subjects/subjects.controller.js]: It's small to facilitate tests, I will change it in the final version, thanks for the reminder o/

File: cg/src/domains/management/subjects/subjects.service.js [hunk 6]
```diff
@@ -2,33 +2,55 @@ const { db } = require('../../../db');
...
+    if (requestedPage == undefined) {
+      requestedPage = 1;
+    }
+    const [numberOfsubjectsObject] = await this.db('subjects')
+      .join('subject_keys', 'subjects.id', '=', 'subject_keys.subject_id')
+      .whereNotNull('personal_data')
+      .whereNotNull('key')
+      .count('personal_data');
+
+    var numberOfsubjects = numberOfsubjectsObject.count;
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.service.js]: we dont use var (the linter actually goes off I think. Or it should I think.)

**Comment by Jakedavies** [comment]
I just had a small comment about using var, after you fix that this looks good to me!

File: cg/src/domains/management/subjects/subjects.service.js (hunk 6 above)
Comment (maksymilian-majer) [reply cg/src/domains/management/subjects/subjects.service.js]: there are pre-commit hooks when committing from root project directory

**Pull request merged**